### PR TITLE
Chibios SPI driver: allow some SPI pins to be left unassigned

### PR DESCRIPTION
**Pull request by purdeaandrei**
## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by purdeaandrei** [comment]
I don't understand why lint is complaining, what it suggests makes no sense to me. I didn't even touch those lines.

image

File: platforms/chibios/drivers/spi_master.c [hunk 1]
```diff
@@ -18,12 +18,20 @@
 
 #include "timer.h"
 
-static pin_t currentSlavePin = NO_PIN;
+static bool spiStarted = false;
```
Comment (sigprof) [inline platforms/chibios/drivers/spi_master.c:21]:
```suggestion
static bool spiStarted = false;

#if SPI_SELECT_MODE == SPI_SELECT_MODE_NONE
static pin_t currentSlavePin;
#endif
```

File: platforms/chibios/drivers/spi_master.c [hunk 2]
```diff
@@ -18,12 +18,20 @@
...
+#    else
 static SPIConfig spiConfig = {false, NULL, 0, 0, 0, 0};
+#    endif
+#else
+#    if defined(K20x) || defined(KL2x) || defined(RP2040)
+static SPIConfig spiConfig = {NULL, 0};
+#    else
+static SPIConfig spiConfig = {false, NULL, 0, 0};
+#    endif
 #endif
```
Comment (sigprof) [inline platforms/chibios/drivers/spi_master.c:35]: Can we just rely on the default zero initialization for `static` variables here?
```c
static SPIConfig spiConfig;
```
(GitHub does not allow adding a proper suggestion for this code block.)

File: platforms/chibios/drivers/spi_master.c [hunk 3]
```diff
@@ -33,26 +41,42 @@ __attribute__((weak)) void spi_init(void) {
...
 
 bool spi_start(pin_t slavePin, bool lsbFirst, uint8_t mode, uint16_t divisor) {
-    if (currentSlavePin != NO_PIN || slavePin == NO_PIN) {
+#if SPI_SELECT_MODE == SPI_SELECT_MODE_NONE
+    if (spiStarted) {
+#else
+    if (spiStarted || slavePin == NO_PIN) {
+#endif
         return false;
     }
```
Comment (sigprof) [inline platforms/chibios/drivers/spi_master.c:81]: Maybe this block breaks `clang-format`?  Then you may try rewriting it in a way that does not have unbalanced braces inside `#if` blocks:
```suggestion
    if (spiStarted) {
        return false;
    }
#if SPI_SELECT_MODE != SPI_SELECT_MODE_NONE
    if (slavePin == NO_PIN) {
        return false;
    }
#endif
```

File: platforms/chibios/drivers/spi_master.c [hunk 4]
```diff
@@ -247,11 +271,15 @@ bool spi_start(pin_t slavePin, bool lsbFirst, uint8_t mode, uint16_t divisor) {
...
 #endif
 
-    currentSlavePin  = slavePin;
+    spiStarted = true;
```
Comment (sigprof) [inline platforms/chibios/drivers/spi_master.c:274]:
```suggestion
    spiStarted = true;
#if SPI_SELECT_MODE == SPI_SELECT_MODE_NONE
    currentSlavePin = slavePin;
#endif
```

File: platforms/chibios/drivers/spi_master.c [hunk 5]
```diff
@@ -247,11 +271,15 @@ bool spi_start(pin_t slavePin, bool lsbFirst, uint8_t mode, uint16_t divisor) {
...
+#if SPI_SELECT_MODE == SPI_SELECT_MODE_PAD
     spiConfig.ssport = PAL_PORT(slavePin);
     spiConfig.sspad  = PAL_PAD(slavePin);
-
     setPinOutput(slavePin);
+#elif SPI_SELECT_MODE != SPI_SELECT_MODE_NONE
```
Comment (sigprof) [inline platforms/chibios/drivers/spi_master.c:279]:
```suggestion
#elif SPI_SELECT_MODE == SPI_SELECT_MODE_NONE
    if (slavePin != NO_PIN) {
        setPinOutput(slavePin);
    }
#else
```

File: platforms/chibios/drivers/spi_master.c [hunk 6]
```diff
@@ -247,11 +271,15 @@ bool spi_start(pin_t slavePin, bool lsbFirst, uint8_t mode, uint16_t divisor) {
...
+#    error "Unsupported SPI_SELECT_MODE"
+#endif
+
     spiStart(&SPI_DRIVER, &spiConfig);
     spiSelect(&SPI_DRIVER);
```
Comment (sigprof) [inline platforms/chibios/drivers/spi_master.c:284]:
```suggestion
    spiSelect(&SPI_DRIVER);
#if SPI_SELECT_MODE == SPI_SELECT_MODE_NONE
    if (slavePin != NO_PIN) {
        writePinLow(slavePin);
    }
#endif
```

File: platforms/chibios/drivers/spi_master.c [hunk 7]
```diff
@@ -283,9 +311,9 @@ spi_status_t spi_receive(uint8_t *data, uint16_t length) {
 }
 
 void spi_stop(void) {
-    if (currentSlavePin != NO_PIN) {
+    if (spiStarted) {
         spiUnselect(&SPI_DRIVER);
```
Comment (sigprof) [inline platforms/chibios/drivers/spi_master.c:315]:
```suggestion
#if SPI_SELECT_MODE == SPI_SELECT_MODE_NONE
        if (currentSlavePin != NO_PIN) {
            writePinHigh(currentSlavePin);
        }
#endif
        spiUnselect(&SPI_DRIVER);
```
(this is the only place where `currentSlavePin` is needed)

**Review by sigprof — COMMENTED** [review]
I had some different ideas in mind when suggesting to add the `NO_PIN` support to the SPI driver; the suggested changes are mostly about that (except one block that might fix the formatting issues).

File: platforms/chibios/drivers/spi_master.c (hunk 2 above)
Comment (purdeaandrei) [reply platforms/chibios/drivers/spi_master.c]: DONE

**Comment by purdeaandrei** [comment]
@sigprof  I applied all your suggestions and rerun with my keyboard and things work.

**Comment by elpekenin** [comment]
Im no big expert but i would get rid of `#if SPI_SELECT_MODE == SPI_SELECT_MODE_NONE` checks around  `writePin{Low/High}`.
- Setting the pin low/high manually, even if ChibiOS also does it internally, shouldn't be a real performance difference and code is slightly easier to read (imo)
- Removing those checks (and maybe some others in the file too) -or moving them to runtime- is probably a good idea for multi-bus SPI support and/or dynamic configuration of the bus(es) if such things ever get added.

PS: This addition is great as it should be used by #20258 if it gets somewhere. Since CS ***could*** be a "virtual pin" (instead of completely unused), ChibiOS wouldn't be able to control it, so we would want:
- `spiSelectI` and `spiUnselectI` (macros that control CS) to become no-ops (using `SPI_SELECT_MODE_NONE`)
- `writePin{Low/High}` to control te "pin" (no matter what it actually is) ourselves and ensure it has the correct value

**Comment by purdeaandrei** [comment]
@elpekenin I don't think ChibiOS currently supports configuring some SPI buses with SPI_SELECT_MODE_NONE and some others with SPI_SELECT_MODE_PAD, correct me if I'm wrong. So not sure what advantage it would have to move those checks to runtime. However if this changes in the future then absolutely yes.

I don't think the current PR implementation would prevent it from working with 20258.

**Comment by elpekenin** [comment]
>  not sure what advantage it would have 

Indeed, it isnt an option currently, as `spiSelectI` gets expanded to different things based on the selected mode.  That's why i said runtime would allow such thing, we could have `switch (bus_config.select_mode)` to use the different things that the macro expands to... However im not sure how different those are from each other, maybe the addition isnt very useful

> I don't think the current PR implementation would prevent it from working with 20258.

It wouldnt, in fact it helps (in case someone tries to use virtual pins for MISO/MOSI, we can ignore those as it makes no sense... and virtual CS could be used thanks to the manual control of the pin over letting ChibiOS handle that)

**Comment by purdeaandrei** [comment]
> Indeed, it isnt an option currently, as spiSelectI gets expanded to different things based on the selected mode. That's why i said runtime would allow such thing, we could have switch (bus_config.select_mode) to use the different things that the macro expands to... However im not sure how different those are from each other, maybe the addition isnt very useful

I like the idea of having this all be runtime selected if that becomes an option in the future in Chibios upstream. But i guess I don't see the point in writing code for it now, since we don't even know if it's every gonna happen, and we don't know what it's gonna look like. If it becomes possible we can always switch to runtime later as a response to the upstream change.

And just going from `#if SPI_SELECT_MODE == SPI_SELECT_MODE_NONE` to `if (SPI_SELECT_MODE == SPI_SELECT_MODE_NONE)` doesn't make sense to me either, since pretty much everywhere in chibios code the checks are done at the preprocessor level. And some of these decisions cannot even be made at runtime, because depending on `SPI_SELECT_MODE`, the `spiConfig` structure gets different fields, so the following decision cannot be made runtime:
```c
#if SPI_SELECT_MODE == SPI_SELECT_MODE_PAD
    spiConfig.ssport = PAL_PORT(slavePin);
    spiConfig.sspad  = PAL_PAD(slavePin);
```

**Review by sigprof — APPROVED** [review]
The ChibiOS code looks like it should work.

One potential issue is that the AVR SPI driver might eventually need a similar change to make it accept `NO_PIN`, but I'm not sure about any actual users for that feature.

**Comment by drashna** [comment]
This has a merge conflict that needs resolving.

**Review by tzarc — APPROVED** [review]

**Pull request merged**